File: test/lua/mdsim/particle_groups/region_species.lua
--
-- Copyright © 2020      Felix Höfling
-- Copyright © 2014-2015 Nicolas Höft
--
-- This file is part of HALMD.
--
-- HALMD is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Lesser General Public License as
-- published by the Free Software Foundation, either version 3 of
-- the License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Lesser General Public License for more details.
--
-- You should have received a copy of the GNU Lesser General
-- Public License along with this program.  If not, see
-- <http://www.gnu.org/licenses/>.
--

local mdsim = halmd.mdsim
local observables = halmd.observables
local random = halmd.random
local writers = halmd.io.writers

local function setup(args)
    local dimension = args.dimension      -- dimension of space
    local np = args.particles             -- number of particles

    local length = {}
    for i = 1, dimension do
        length[i] = 10
    end
    -- create simulation domain with periodic boundary conditions
    local box = mdsim.box({length = length})

    -- create system state for all particles first
    local particle = mdsim.particle({dimension = dimension, particles = np, species = 2})

    -- set particle species, with continuous range of IDs per species
    local species = {}
    for i = 1, np do
        table.insert(species, (i > np / 2) and 1 or 0)
    end
    particle.data["species"] = species

    -- set initial particle positions
    mdsim.positions.lattice({box = box, particle = particle}):set()
    -- randomly shuffle the positions
    particle.data["position"] = random.generator({memory = "host"}):shuffle(particle.data["position"])

    if particle.memory == "gpu" then
      -- sort particles *once*, as they do not move
      local sort = mdsim.sorts.hilbert({box = box, particle = particle})
      sort:order()
    end

    -- select particles within/not within upper quadrant:
    -- define geometry first
    local lowest_corner = {}
    for i = 1, dimension do
        lowest_corner[i] = 0
        length[i] = length[i] / 2
    end
    local cuboid = mdsim.geometries.cuboid({lowest_corner = lowest_corner, length = length})

    -- construct included/excluded particle groups, label is inherited from 'region'
    local group = {}
    group["included"] = mdsim.particle_groups.region_species({
        particle = particle, geometry = cuboid, selection = "included"
      , species = 1
      , label = "upper quadrant (included)"
    })
    group["excluded"] = mdsim.particle_groups.region_species({
        particle = particle, geometry = cuboid, selection = "excluded"
      , species = 1
      , label = "upper quadrant (excluded)"
    })

    return group, cuboid, args
end

local function test(group, cuboid, args)
    -- the total number of included/excluded particles must
    -- equal the number of particles of species 1
    print(("group sizes: %d (inc) + %d (exc) = %d")
       :format(group["included"].size, group["excluded"].size, args.particles / 2)
    )
    assert(group["included"].size + group["excluded"].size == args.particles / 2)

    local lowest_corner = cuboid.lowest_corner
    local length = cuboid.length

    -- convert to a new particle instance in order to access the positions
    local particle_exc = group["excluded"]:to_particle()
    local particle_inc = group["included"]:to_particle()

    -- for included/excluded check that the particles have been sorted
    -- into the respective group correctly
    local positions_inc = particle_inc.data.position
    local species_inc = particle_inc.data.species
    for i = 1, group["included"].size do
        assert(species_inc[i] == 1, ("particle #%d of species %d wrongly selected"):format(i, species_inc[i]))
        local r = positions_inc[i]
        for j = 1, #r do
            local dr = r[j] - lowest_corner[j]
            assert(dr < length[j] and dr > 0, ("particle #%d wrongly included in selection"):format(i))
        end
    end

    local positions_exc = particle_exc.data.position
    local species_exc = particle_exc.data.species
    for i = 1, group["excluded"].size do
        assert(species_exc[i] == 1, ("particle #%d of species %d wrongly selected"):format(i, species_exc[i]))
        local r = positions_exc[i]
        local outside = false
        for j = 1, #r do
            local dr = r[j] - lowest_corner[j]
            if dr > length[j] or dr < 0 then
                outside = true
            end
        end
        assert(outside, ("particle #%d wrongly excluded from selection"):format(i))
    end
end

--
-- Parse command-line arguments.
--
function define_args(parser)
    parser:add_argument("output,o", {type = "string", default = "region_test", help = "prefix of output files"})

    parser:add_argument("particles", {type = "number", default = 10000, help = "number of particles"})
    parser:add_argument("dimension", {type = "number", default = 3, help = "dimension of space"})
end

--
-- set up system and perform test
--
function main(args)
    test(setup(args))
end

File: halmd/mdsim/gpu/particle_groups/region_species.cpp
/*
 * Copyright © 2019 Roya Ebrahimi Viand
 * Copyright © 2014-2015 Nicolas Höft
 *
 * This file is part of HALMD.
 *
 * HALMD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 */

#include <halmd/algorithm/gpu/copy_if.hpp>
#include <halmd/algorithm/gpu/radix_sort.hpp>
#include <halmd/mdsim/geometries/cuboid.hpp>
#include <halmd/mdsim/geometries/cylinder.hpp>
#include <halmd/mdsim/geometries/sphere.hpp>
#include <halmd/mdsim/gpu/particle.hpp>
#include <halmd/mdsim/gpu/particle_groups/region_species.hpp>
#include <halmd/mdsim/gpu/particle_groups/region_species_kernel.hpp>
#include <halmd/utility/lua/lua.hpp>

#include <cuda_wrapper/cuda_wrapper.hpp>
#include <exception>
#include <stdexcept>

namespace halmd {
namespace mdsim {
namespace gpu {
namespace particle_groups {

template <int dimension, typename float_type, typename geometry_type>
region_species<dimension, float_type, geometry_type>::region_species(
    std::shared_ptr<particle_type const> particle
  , std::shared_ptr<geometry_type const> geometry
  , geometry_selection geometry_sel
  , unsigned int species
  , std::shared_ptr<logger> logger
)
  : particle_(particle)
  , geometry_(geometry)
  , geometry_selection_(geometry_sel)
  , species_(species)
  , logger_(logger)
{
    geometry_->log(logger_);
    try {
        auto mask = make_cache_mutable(mask_);
        auto selection = make_cache_mutable(selection_);
        mask->reserve(particle_->dim().threads());
        mask->resize(particle_->nparticle());
        selection->reserve(particle_->dim().threads());
    }
    catch (cuda::error const&) {
        LOG_ERROR("failed to allocate global device memory");
        throw;
    }
}
template <int dimension, typename float_type, typename geometry_type>
cache<typename region_species<dimension, float_type, geometry_type>::array_type> const&
region_species<dimension, float_type, geometry_type>::selection()
{
    update_selection_();
    return selection_;
}

template <int dimension, typename float_type, typename geometry_type>
cache<typename region_species<dimension, float_type, geometry_type>::array_type> const&
region_species<dimension, float_type, geometry_type>::mask()
{
    update_mask_();
    return mask_;
}

/**
 * update the particle lists for the region
 */
template <int dimension, typename float_type, typename geometry_type>
void region_species<dimension, float_type, geometry_type>::update_mask_()
{
    cache<position_array_type> const& position_cache = particle_->position();
    if (position_cache != mask_cache_) {
        auto const& position = read_cache(position_cache);

        LOG_DEBUG("update selection mask for region and species");
        scoped_timer_type timer(runtime_.update_mask);

        auto mask = make_cache_mutable(mask_);

        auto& kernel = region_species_wrapper<dimension, geometry_type>::kernel;
        // calculate "bin", ie. inside/outside the region
        cuda::memset((*mask).begin(), (*mask).end(), 0xFF);
        kernel.compute_mask.configure(particle_->dim().grid, particle_->dim().block);
        kernel.compute_mask(
            position.data()
          , particle_->nparticle()
          , mask->data()
          , *geometry_
          , geometry_selection_ == excluded ? particle_groups::excluded : particle_groups::included
          , species_
        );
        mask_cache_ = position_cache;
    }
}

/**
 * update the particle lists for the region
 */
template <int dimension, typename float_type, typename geometry_type>
void region_species<dimension, float_type, geometry_type>::update_selection_()
{
    auto const& position_cache = particle_->position();

    if(position_cache != selection_cache_) {
        auto const& position = read_cache(position_cache);
        unsigned int nparticle = particle_->nparticle();

        LOG_DEBUG("update particle selection for region and species");
        scoped_timer_type timer(runtime_.update_selection);

        auto selection = make_cache_mutable(selection_);
        auto size = make_cache_mutable(size_);

        // allocate memory for maximum selection (i.e., all particles)
        selection->resize(nparticle);

        auto const& kernel = region_species_wrapper<dimension, geometry_type>::kernel;
        unsigned int s = kernel.copy_selection(
            position.data()
          , nparticle
          , selection->data()
          , *geometry_
          , geometry_selection_ == excluded ? particle_groups::excluded : particle_groups::included
          , species_
        );
        // shrink array to actual size of selection
        selection->resize(s);
        *size = s;

        selection_cache_ = position_cache;
    }
}

template <int dimension, typename float_type, typename geometry_type>
cache<typename region_species<dimension, float_type, geometry_type>::array_type> const&
region_species<dimension, float_type, geometry_type>::ordered()  // ID order
{
    auto const& s = selection();
    if (s != ordered_cache_) {
//        auto const& id = read_cache(particle_->id());

        LOG_WARNING("sorting selection of particle indices not yet implemented");
//        LOG_DEBUG("sorting selection of particle indices");
        scoped_timer_type timer(runtime_.sort_selection);

        auto ordered = make_cache_mutable(ordered_);
        ordered->resize(s->size());
        cuda::copy(s->begin(), s->end(), ordered->begin());

        // TODO: bring selection in ID order, sort 'ordered' by key 'id'
        // 1) copy IDs of the selection, 2) perform in-place sort
        // radix_sort(key.begin(), key.end(), ordered->begin());

        ordered_cache_ = s;
    }
    return ordered_;
}

template <int dimension, typename float_type, typename geometry_type>
cache<typename region_species<dimension, float_type, geometry_type>::array_type> const&
region_species<dimension, float_type, geometry_type>::unordered()  // memory order
{
    return selection();
}

template <int dimension, typename float_type, typename geometry_type>
cache<typename region_species<dimension, float_type, geometry_type>::size_type> const&
region_species<dimension, float_type, geometry_type>::size()
{
    update_selection_();
    return size_;
}

template <typename particle_group_type, typename particle_type>
static void wrap_to_particle(
    std::shared_ptr<particle_group_type> self
  , std::shared_ptr<particle_type> particle_src
  , std::shared_ptr<particle_type> particle_dst
)
{
    particle_group_to_particle(*particle_src, *self, *particle_dst);
}

template <int dimension, typename float_type, typename geometry_type>
void region_species<dimension, float_type, geometry_type>::luaopen(lua_State* L)
{
    using namespace luaponte;
    module(L, "libhalmd")
    [
        namespace_("mdsim")
        [
            namespace_("particle_groups")
            [
                class_<region_species, particle_group>()
                    .scope
                    [
                        class_<runtime>("runtime")
                            .def_readonly("update_mask", &runtime::update_mask)
                            .def_readonly("update_selection", &runtime::update_selection)
                            .def_readonly("sort_selection", &runtime::sort_selection)
                    ]
                    .def_readonly("runtime", &region_species::runtime_)
                    .def("to_particle", &wrap_to_particle<region_species<dimension, float_type, geometry_type>, particle_type>)

              , def("region_species", &std::make_shared<region_species<dimension, float_type, geometry_type>
                  , std::shared_ptr<particle_type const>
                  , std::shared_ptr<geometry_type const>
                  , geometry_selection
                  , unsigned int
                  , std::shared_ptr<logger>
                >)
            ]
        ]
    ];
}

HALMD_LUA_API int luaopen_libhalmd_mdsim_gpu_particle_groups_region_species(lua_State* L)
{
#ifdef USE_GPU_SINGLE_PRECISION
    region_species<3, float, mdsim::geometries::cuboid<3, float>>::luaopen(L);
    region_species<2, float, mdsim::geometries::cuboid<2, float>>::luaopen(L);
    region_species<3, float, mdsim::geometries::cylinder<3, float>>::luaopen(L);
    region_species<2, float, mdsim::geometries::cylinder<2, float>>::luaopen(L);
    region_species<3, float, mdsim::geometries::sphere<3, float>>::luaopen(L);
    region_species<2, float, mdsim::geometries::sphere<2, float>>::luaopen(L);
#endif
#ifdef USE_GPU_DOUBLE_SINGLE_PRECISION
    region_species<3, dsfloat, mdsim::geometries::cuboid<3, float>>::luaopen(L);
    region_species<2, dsfloat, mdsim::geometries::cuboid<2, float>>::luaopen(L);
    region_species<3, dsfloat, mdsim::geometries::cylinder<3, float>>::luaopen(L);
    region_species<2, dsfloat, mdsim::geometries::cylinder<2, float>>::luaopen(L);
    region_species<3, dsfloat, mdsim::geometries::sphere<3, float>>::luaopen(L);
    region_species<2, dsfloat, mdsim::geometries::sphere<2, float>>::luaopen(L);
#endif
    return 0;
}

// explicit instantiation
#ifdef USE_GPU_SINGLE_PRECISION
template class region_species<3, float, mdsim::geometries::cuboid<3, float>>;
template class region_species<2, float, mdsim::geometries::cuboid<2, float>>;
template class region_species<3, float, mdsim::geometries::cylinder<3, float>>;
template class region_species<2, float, mdsim::geometries::cylinder<2, float>>;
template class region_species<3, float, mdsim::geometries::sphere<3, float>>;
template class region_species<2, float, mdsim::geometries::sphere<2, float>>;
#endif
#ifdef USE_GPU_DOUBLE_SINGLE_PRECISION
template class region_species<3, dsfloat, mdsim::geometries::cuboid<3, float>>;
template class region_species<2, dsfloat, mdsim::geometries::cuboid<2, float>>;
template class region_species<3, dsfloat, mdsim::geometries::cylinder<3, float>>;
template class region_species<2, dsfloat, mdsim::geometries::cylinder<2, float>>;
template class region_species<3, dsfloat, mdsim::geometries::sphere<3, float>>;
template class region_species<2, dsfloat, mdsim::geometries::sphere<2, float>>;
#endif

} // namespace particle_groups
} // namespace gpu
} // namespace mdsim
} // namespace halmd
